### PR TITLE
Added SAC

### PR DESCRIPTION
**Pull request by Veyron2121**
Implementation from https://github.com/denisyarats/pytorch_sac. Not sure how to cite.

### CODE REVIEW

File: rl_games/algos_torch/network_builder.py [hunk 1]
```diff
@@ -761,6 +763,212 @@ def build(self, name, **kwargs):
         net = A2CResnetBuilder.Network(self.params, **kwargs)
         return net
 
+class TanhTransform(pyd.transforms.Transform):
```
Comment (Denys88) [inline rl_games/algos_torch/network_builder.py:766]: Lets put it to the torch_ext.py or create your own sac_helper.py file.

File: rl_games/common/experience.py [hunk 2]
```diff
@@ -194,3 +195,98 @@ def update_priorities(self, idxes, priorities):
 
             self._max_priority = max(self._max_priority, priority)
 
+
+class VectorizedReplayBuffer:
```
Comment (Denys88) [inline rl_games/common/experience.py:199]: I  think you can use pytorch tensors everywhere instead of numpy.
Here are bonuses:
1) you can put everything on gpu
2) no need to cast to the torch tensor

**Review by Denys88 — COMMENTED** [review]
Overall looks good.
I need  take a look a little bit deeper into yaml configs and LGTM.

File: rl_games/algos_torch/SACAgent.py [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+        total_time = 0
+        rep_count = 0
+        self.frame = 0
+        obs = self.env_reset()
+        # self.curr_frames = self.batch_size_envs
+
+        while self.step < self.max_epochs:       
+            # sample action for data collection
+            if self.step < self.num_seed_steps:
+                action = torch.empty(self.num_actors, *self.env_info["action_space"].shape).to(self.sac_device)
```
Comment (Denys88) [inline rl_games/algos_torch/SACAgent.py:375]: you can pass device into the torch.empty. it will be faster.

File: rl_games/algos_torch/SACAgent.py [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+
+
+            if isinstance(obs, dict):
+                obs = obs['obs']
+            if isinstance(next_obs, dict):    
+                next_obs = next_obs['obs']
+
+            # break
+           
+            self.replay_buffer.add(self._to_2d_np_cpu(obs).T, self._to_2d_np_cpu(action).T, self._to_2d_np_cpu(rewards), self._to_2d_np_cpu(next_obs).T, self._to_2d_np_cpu(dones.bool()), self._to_2d_np_cpu(dones_no_max.bool()))
```
Comment (Denys88) [inline rl_games/algos_torch/SACAgent.py:423]: Do you need to cast it to the numpy if you are using Torch Tensor Experience?

File: rl_games/algos_torch/SACAgent.py [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+        self.model.sac_network.critic_target.load_state_dict(weights['critic_target'])
+
+        if self.normalize_input:
+            self.running_mean_std.load_state_dict(weights['running_mean_std'])
+
+    def set_full_state_weights(self, weights):
+        self.set_weights(weights)
+
+        self.step = weights['step']
+        if self.has_central_value:
```
Comment (Denys88) [inline rl_games/algos_torch/SACAgent.py:219]: you can remove unused networks like rnd or asymmetric

File: rl_games/algos_torch/SACAgent.py [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+        # self.writer = SummaryWriter('ant_runs/' + config['name'] + datetime.now().strftime("_%d-%H-%M-%S"))
+        self.writer = SummaryWriter('tested_new/' + 'ant')
+        # self.writer = SummaryWriter('walker/'+'fixed_buffer')
+        print("Run Directory:", config['name'] + datetime.now().strftime("_%d-%H-%M-%S"))
+
+        self.is_tensor_obses = None
+
+        self.curiosity_config = self.config.get('rnd_config', None)
+        self.has_curiosity = self.curiosity_config is not None
+        if self.has_curiosity:
```
Comment (Denys88) [inline rl_games/algos_torch/SACAgent.py:155]: I removed curiosity support because I've found that rnd curiosity works the best only for some atari games. But exploration based reward tuned for special case performs much better.

**Review by Denys88 — APPROVED** [review]
Please fix a few comments. And make one more pass by removing unused functionality.
I've approved it. Feel free to push this version and create additional prs with improvements.
